Search preview version 2024-03-01-preview

sazeesha098 opened the pull request:
# Data Plane API - Pull Request
Another BYOV preview 

## API Info: The Basics
Target Features:

* Support narrow data types for vector values and vector quantization
* Support reduction of vector field storage overhead via _source exclusions.

sazeesha committed:
baseline for preview search 2024-03-01-preview

sazeesha committed:
Updating the error response

sazeesha committed:
Update readme.md with new preview version

jhendrixMSFT commented:
CC @JeffreyRichter for the breaking change review.

sazeesha committed:
Support narrow data types

sazeesha committed:
vectors add stored property

JeffreyRichter commented:
Why the new "required" properties? These are breaking and should be avoided.

sazeesha098 commented:
> Why the new "required" properties? These are breaking and should be avoided.

Similar to other Azure services, we had to encapsulate the SearchError within an Error property. While our service successfully accomplishes this, Swagger fails to include it. For instance, the service returns:

{
  "error": {
    "message": "<required string>",
    "code": "<optional string>",
    "details": "<optional array of SearchError (this type)>"
  }
}

However, Swagger only exhibits the following:
{
  "message": "<required string>",
  "code": "<optional string>",
  "details": "<optional array of SearchError (this type)>"
}

JeffreyRichter commented:
OK, I approved it

sazeesha committed:
SpellCheck override for rerank

sazeesha committed:
Update Support narrow data types

sazeesha committed:
update descriptions

sazeesha committed:
Update SpellCheck override

sazeesha committed:
Swagger PrettierCheck

heaths commented on specification/search/data-plane/Azure.Search/readme.md:
````diff
@@ -171,6 +171,45 @@ directive:
+- from: preview/2024-03-01-Preview/searchservice.json
+  where: $.definitions.SearchError
+  transform: $["x-ms-client-name"] = "SearchServiceError"
+```
````
Yes, that's ideal. Model names don't appear in the JSON payload, so it's preferred to change these as long as they do not break existing SDK code.

sazeesha098 replied:
Updated it like other Azure services that use ErrorResponse object from common-types

heaths commented on specification/search/data-plane/Azure.Search/preview/2024-03-01-Preview/searchindex.json:
```diff
@@ -1868,6 +1868,11 @@
           "type": "boolean",
           "description": "When true, triggers an exhaustive k-nearest neighbor search across all vectors within the vector index. Useful for scenarios where exact matches are critical, such as determining ground truth values."
+        },
+        "oversampling": {
```
Within this domain, is this really treated as a single word, or should it be `overSampling`? I defer to your team's expertise, of course. Just wanted to raise the question.

robertklee replied:
thanks for the callout. I double checked, it is one term in the industry. Originated in signal processing.

![image](https://github.com/Azure/azure-rest-api-specs/assets/32447910/2e8a8e51-98b1-47ca-91d1-3f51c04c3be3)

heaths commented on specification/search/data-plane/Azure.Search/preview/2024-03-01-Preview/searchservice.json:
```diff
@@ -6184,6 +6209,12 @@
+        "compression": {
+          "x-ms-client-name": "compressionConfigurationName",
+          "type": "string",
+          "description": "The name of the compression method configuration that specifies the compression method and optional parameters.",
```
This makes it sound like it's a discriminated field. Is that the case? Are the possible values known? What would customers do with this?

heaths replied:
I see below there's a `name` property, but why not just have this property use `$ref` to `VectorSearchCompressionConfiguration`? What you have doesn't produce a type-safe and compiled-checked SDK.

robertklee replied:
the profiles object aggregates all the vector features a customer may want to bundle together. So in profiles, they can reference algorithm config named "my-custom-hnsw", add a compression config they named "my-scalar-quantization", etc.

Here in the profiles object, they reference the name they've set, so it's not a discriminated field because it's customer defined.

we use this pattern for custom analyzers/normalizers as well.

heaths commented on specification/search/data-plane/Azure.Search/preview/2024-03-01-Preview/searchservice.json:
```diff
@@ -6352,6 +6383,105 @@
+      },
+      "description": "Contains the parameters specific to Scalar Quantization."
+    },
+    "VectorSearchCompressionTargetType": {
```
Why is this "type" and others are "kind"? We prefer "kind".

robertklee replied:
good point. i support renaming to 'kind' suffix as well

robertklee replied:
Feedback from API review: rename to "VectorSearchCompressionTargetKind"

robertklee replied:
Reflecting on this a bit, I propose the name "VectorSearchCompressionTarget**Data**Type". Kind sounds a bit weird in this scenario because it is describing possible data _types_, which is a standard terminology in the software field.

robertklee replied:
done

heaths reviewed (approved):
Since we discussed some of my questions in the meeting, I'll sign off for a preview. Consider other changes as needed, though.

BevLoh commented on specification/search/data-plane/Azure.Search/preview/2024-03-01-Preview/searchservice.json:
```diff
@@ -3145,7 +3145,10 @@
         "Edm.ComplexType",
-        "Edm.Single"
+        "Edm.Single",
+        "Edm.Float16",
```
Feedback from API review: consider naming it Edm.Half

sazeesha098 replied:
Updated to use Edm.Half

BevLoh commented on specification/search/data-plane/Azure.Search/preview/2024-03-01-Preview/searchservice.json:
```diff
@@ -6352,6 +6383,105 @@
+      },
+      "description": "Contains configuration options specific to the scalar quantization compression method used during indexing and querying."
+    },
+    "SqParameters": {
```
Feedback from API review: Rename to ScalarQuantizationParameters

robertklee replied:
done

Commit:
api review feedback

Commit:
update samples

Commit:
update float16 -> half

sazeesha committed:
Update examples with renamed properties

sazeesha committed:
Use ErrorResponse from common-types

BevLoh commented on specification/search/data-plane/Azure.Search/readme.md:
````diff
@@ -171,6 +167,34 @@ directive:
+- preview/2024-03-01-Preview/searchindex.json
+```
+
+### Tag: package-2023-10-searchservice-preview
````
what are the 2023-10-* tags for?

sazeesha098 replied:
Updated it correctly to 2024-03-* tags

sazeesha committed:
Fix the Tag: package-2024-03 in readme

sazeesha098 commented:
@heaths, thanks for all the feedback, we have addressed it and updated this PR. Can you please add the "APIStewardshipBoard-SignedOff" tag. 

cc: @jhendrixMSFT

jhendrixMSFT commented:
I think @JeffreyRichter has the final call for adding `APIStewardshipBoard-SignedOff`

robertklee commented on specification/search/data-plane/Azure.Search/preview/2024-03-01-Preview/examples/SearchIndexSearchDocumentsPost.json:
```diff
@@ -110,7 +111,7 @@
           }
         ],
-        "@odata.nextLink": "https://myservice.search.windows.net/indexes('myindex')/docs/search.post.search?api-version=2024-03-01-Preview"
+        "@odata.nextLink": "https://myservice.search.windows.net/indexes('myindex')/docs/search.post.search?api-version=2023-10-01-Preview"
```
nit, this api version is wrong i think

```suggestion
        "@odata.nextLink": "https://myservice.search.windows.net/indexes('myindex')/docs/search.post.search?api-version=2024-03-01-Preview"
```

robertklee commented on specification/search/data-plane/Azure.Search/preview/2024-03-01-Preview/searchindex.json:
```diff
@@ -1868,6 +1868,11 @@
+        "oversampling": {
+          "type": "number",
+          "format": "double",
+          "description": "Oversampling factor. Minimum value is 1. It overrides the defaultOversampling parameter configured in the index definition. It can be set only when 'rerankWithOriginalVectors' is true. This parameter is only permitted when a compression method is used on the underlying vector field."
```
```suggestion
          "description": "Oversampling factor. Minimum value is 1. It overrides the 'defaultOversampling' parameter configured in the index definition. It can be set only when 'rerankWithOriginalVectors' is true. This parameter is only permitted when a compression method is used on the underlying vector field."
```

robertklee reviewed (approved):
minor nits. LGTM

sazeesha committed:
Address comments

sazeesha098 commented:
> I think @JeffreyRichter has the final call for adding `APIStewardshipBoard-SignedOff`

sazeesha098 commented:
@jhendrixMSFT can you please merge this PR?

HeidiSteen commented on specification/search/data-plane/Azure.Search/preview/2024-03-01-Preview/examples/SearchServiceGetIndexerStatus.json:
```diff
@@ -0,0 +1,89 @@
+        "lastResult": {
+          "status": "success",
+          "errorMessage": null,
+          "startTime": "2014-11-26T03:37:18.853Z",
```
2024

HeidiSteen commented on specification/search/data-plane/Azure.Search/preview/2024-03-01-Preview/examples/SearchServiceGetIndexerStatus.json:
```diff
@@ -0,0 +1,89 @@
+          {
+            "status": "success",
+            "errorMessage": null,
+            "startTime": "2014-11-26T03:37:18.853Z",
```
2024

HeidiSteen commented on specification/search/data-plane/Azure.Search/preview/2024-03-01-Preview/examples/SearchServiceGetIndexerStatus.json:
```diff
@@ -0,0 +1,89 @@
+          {
+            "status": "transientFailure",
+            "errorMessage": null,
+            "startTime": "2014-11-26T03:28:10.125Z",
```
2024

HeidiSteen commented on specification/search/data-plane/Azure.Search/preview/2024-03-01-Preview/searchindex.json:
```diff
@@ -0,0 +1,2826 @@
+          }
+        ]
+      },
+      "description": "The language of the query."
```
This parameter exists for backwards compatibility. In this API version, it's used for speller only, and speller is available in a subset of these languages.

HeidiSteen reviewed (commented):
Left some review comments -- not sure if you can take it, but would appreciate it if you tried.
